### ISSUE TITLE
PageRank algo convergence condition

### ISSUE THREAD

**Issue by greghogan** [networkx/networkx#4792]
This issue was raised in #1465 six years ago but incorrectly dismissed. The tolerance should not be multiplied by the number of vertices since the sum of vertex scores is `1` irrespective of the graph's vertex count.

For large graphs of around one million vertices the current implementation will likely terminate after a single iteration.

### Current Behavior
For the example below, the convergence error for the current PageRank implementation is `err = 0.322`.

### Expected Behavior
The convergence error should be on the order of the tolerance, which by default is `1e-06`. By dividing `tol` by `N` we undo the multiplication by `N` in PageRank and generate a convergence error (L1-norm) of `err_accurate = 2.23e-07`.

Both of the `pagerank` and `pagerank_scipy` methods erroneously multiply `tol` by `N`.

### Steps to Reproduce
```
import inspect
import networkx as nx

# https://snap.stanford.edu/data/com-Youtube.html
G = nx.read_edgelist('com-youtube.ungraph.txt')
N = G.number_of_edges()

tol = inspect.signature(nx.pagerank).parameters['tol'].default
pr_low_tol = nx.pagerank(G, alpha=0.9, tol=10**-12/N, max_iter=1000)

# will compare "err < N * tol"
pr = nx.pagerank(G, alpha=0.9)
err = sum(abs(pr[i]-pr_low_tol[i]) for i in pr_low_tol.keys())

# will compare "err < N * tol / N" == "err < tol"
pr_accurate = nx.pagerank(G, alpha=0.9, tol=tol/N)
err_accurate = sum(abs(pr_accurate[i]-pr_low_tol[i]) for i in pr_low_tol.keys())

tol, err, err_accurate
```

### Environment
Python version: 3.9.2
NetworkX version: 2.5

**Comment by dschult** [comment]
Convergence criteria are tricky...

Here we are questioning what scale to use for determining convergence of the algorithm. There are many ways to define the norm of the convergence error and many ways to scale that norm to make it meaningful or useful. 

We scale by the number of **nodes** in the graph (not number of edges as in your example, but in this case it doesn't matter much). That is, `tol` is the "per node error tolerance". We choose this style because then a reasonable default works well for any size graph.  As you point out, you can specify any tolerance you'd like so you can un-do this choice of tolerance if you prefer a different tolerance measure.  BUT, you can only do so if you **know how tol is actually used in the function**!

That probably means the best solution is to inform the users and choose a reasonable default.

I propose that we leave the default behavior as it is, but add to the description of the tolerance parameter for the `pagerank` function. 
```
    tol : float, optional
        Error tolerance used to check convergence in power method solver.
```
could be replaced by
```
    tol : float, optional
        Error tolerance used to check convergence in power method solver.
        This is scaled by the number of nodes so that large graphs aren't held
        to an unrealistic standard. That is, `abs(x - xlast).sum() < len(G) * tol`
        is the criteria stopping criteria. If you prefer `tol` to be the vector tolerance
        you should divide your tolerance level by the number of nodes in `G`.
```

Better wording can probably be used.

**Comment by greghogan** [comment]
@dschult thank you for looking into this. Yes, this should have read `N = G.number_of_nodes()`. With this fix we still have `err = 0.322` and now `err_accurate = 5.393e-07` (closer to `tol = 1e-06`).

The effect of using `N * tol` is that for many non-trivial graphs PageRank will terminate after only a single iteration, and it is the responsibility of the user to understand the algorithm and how the size of the graph affects the accuracy of the results.

By instead comparing `err < tol` the algorithm will generate consistently accurate results independent of graph size. With the default `tol=1.0e-6` we understand that the scores will have converged to within one part per million regardless of the number of nodes.

Even with expanded documentation, the existing implementation will return accurate results (reasonably close to `tol`) during testing and deployment with toy graphs but increasingly inaccurate results as the dataset increases in size (this is how we stumbled upon the issue). And there is no mechanism by which users can detect this change since the number of iterations is not reported. We should instead prefer reasonable defaults that work well for any size graph.

**Comment by dschult** [comment]
Yes, both of us want reasonable defaults that work well for any size graph. The feeling is/was that `N*tol` would do that. You are saying it doesn't (and based on real world graphs and data). This is definitely worth considering/pursuing. 

One difficulty is the phrase "the scores will have converged to one part per million".  If I understand correctly, the current system interprets this as "each score will have converged to..." while your suggestion is closer to "the scores as a whole will have converged to...".

This leads you to compute the error by using the norm instead of the norm divided by the number of elements in the vector.  I would have to re-look up the implications of this distinction. But is that difference clear?  I'm trying to call into question your assumption that the norm is the correct way to measure the error of this calculation.

Said another way, if the current calculation stops after one iteration then the average error of each of the scores is less than one in a million -- as requested. If we use `err<tol` then the average error over all the scores must become very small as the number of nodes increases.  Does this makes sense?

**Comment by greghogan** [comment]
> If we use `err<tol` then the average error over all the scores must become very small as the number of nodes increases.

Yes, because as the number of nodes increases the average score becomes similarly smaller yet the relative error (`err`) remains a consistent measure of convergence.

**Comment by Qudirah** [comment]
> 

Can I make this correction?

**Comment by rossbar** [comment]
> Can I make this correction?

Sure - IMO the parameter description from `tol` could be updated to note that the convergence condition is `N * tol`, since that is an accurate representation of the current behavior.

Note that you'll likely have to update multiple docstrings, as `pagerank` calls the private `_pagerank_scipy` under-the-hood.

**Comment by Peiffap** [comment]
In #6212, a comment was added to the docstring of the `pagerank` function to clarify the ambiguity, but this issue wasn't closed. Is there any particular reason for that? Is additional clarification necessary?
